crates/executor/src/estimate.rs

```diff
@@ -13,54 +19,359 @@ pub fn estimate(
+        Ok(tx_info) => tx_info,
+        Err(TransactionSimulationError::ExecutionError(error)) => {
+            return Err(error);
+        }
+        Err(TransactionSimulationError::OutOfGas) => {
+            unreachable!("Transaction should not run out of gas with MAX gas limit")
+        }
+    };
+
+    let l2_gas_consumed = tx_info.receipt.gas.l2_gas;
+
+    // TODO: g = (1 + EPSILON) * g
+    // This step is missing. Check what EPSILON is, then implement it.
+
+    let mut lower_bound = l2_gas_consumed;
+    let mut upper_bound = GasAmount::MAX;
+
+    let mut current_l2_gas_limit = midpoint(lower_bound, upper_bound);
+
+    // Run a binary search to find the minimal gas limit that still allows the
+    // transaction to execute without running out of L2 gas.
+    loop {
+        tracing::debug!(current_limit=%current_l2_gas_limit, "Searching for minimal L2 gas limit");
+        update_l2_gas_limit(tx, current_l2_gas_limit);
+
+        match simulate_transaction(tx, tx_index, state, block_context) {
+            Ok(_) => {
+                if search_done(lower_bound, upper_bound, L2_GAS_SEARCH_MARGIN) {
+                    break;
+                }
+
+                upper_bound = current_l2_gas_limit;
+                current_l2_gas_limit = midpoint(lower_bound, upper_bound);
+            }
+            Err(TransactionSimulationError::OutOfGas) => {
+                lower_bound = current_l2_gas_limit;
+                current_l2_gas_limit = midpoint(lower_bound, upper_bound);
+            }
+            Err(TransactionSimulationError::ExecutionError(error)) => {
+                return Err(error);
+            }
+        }
+    }
+
+    if current_l2_gas_limit > initial_l2_gas_limit {
+        tracing::debug!(
+            initial_limit=%initial_l2_gas_limit,
+            final_limit=%current_l2_gas_limit,
+            "Initial L2 gas limit exceeded"
+        );
+        // Set the L2 gas limit to zero so that the transaction reverts with a detailed
+        // `ExecutionError`.
+        update_l2_gas_limit(tx, GasAmount::ZERO);
+        match execute_transaction(tx, tx_index, state, block_context) {
+            Err(e @ TransactionExecutionError::ExecutionError { .. }) => {
+                return Err(e);
+            }
+            _ => unreachable!("Transaction should revert when gas limit is zero"),
+        }
+    }
+
+    // Finally, execute the transaction with the found L2 gas limit and set that
+    // limit as the estimate.
+    let mut tx_info = execute_transaction(tx, tx_index, state, block_context)
+        .expect("Transaction already executed successfully");
+    tx_info.receipt.gas.l2_gas = current_l2_gas_limit;
+
+    Ok(tx_info)
+}
+
+/// Calculates the midpoint between two gas amounts without overflowing.
+fn midpoint(a: GasAmount, b: GasAmount) -> GasAmount {
+    let GasAmount(a) = a;
+    let GasAmount(b) = b;
+    GasAmount(a + (b - a) / 2)
```

Comment:
This does not handle u64 overflows. I know it's rare, but still, we should be careful.

Reply:
Yeah, that's a good point, but it's not an obvious requirement if you check the function. What about adding a (debug) assert to document that behavior?